File: brick_wall.go
// Problem: Brick Wall
// (https://leetcode.com/problems/brick-wall/#/description)

func leastBricks(wall [][]int) int {
    width := 0
    holes := make([]map[int]bool, len(wall))
    allHoles := make(map[int]int)
    for index, layer := range wall {
        if holes[index] == nil {
            holes[index] = make(map[int]bool)
        }
        cracks := 0
        for _, brick := range layer[:len(layer) - 1] {
            cracks += brick
            holes[index][cracks] = true
            if _, ok := allHoles[cracks]; !ok {
                allHoles[cracks] = 0
            }
            allHoles[cracks] += 1
        }
        if width == 0 {
            cracks += layer[len(layer) - 1]
            width = cracks
        }
    }
    if len(allHoles) == 0 {
        return len(wall)
    }
    minBricks := 100001
    for _, num := range allHoles {
        bricks := len(wall) - num
        if bricks < minBricks {
            minBricks = bricks
        }
    }
    return minBricks
}

File: brick_wall.py
# Problem: Brick Wall
# (https://leetcode.com/problems/brick-wall/description/)

from collections import defaultdict
class Solution(object):
    def leastBricks(self, wall):
        """
        :type wall: List[List[int]]
        :rtype: int
        """
        if len(wall) == 0 or len(wall[0]) == 0: return 0
        
        width = sum(wall[0])
        slits = defaultdict(int)
        if width == 1: return len(wall)
        
        for layer in wall:
            slit = 0
            for brick in layer:
                slit += brick
                if slit == width: break
                slits[slit] += 1
        return len(wall) - max(slits.values()) if len(slits) != 0 else len(wall)
        